src/operator/aggregate/mod.rs

```diff
@@ -0,0 +1,1057 @@
+    {
+        self.weigh(f).aggregate_generic::<TS, _, _>(WeightedCount)
+    }
+
+    /// Convert indexed Z-set `Z` into a Z-set where the weight of each key
+    /// is computed as:
+    ///
+    /// ```text
+    ///    __
+    ///    ╲
+    ///    ╱ f(k,v) * w
+    ///    ‾‾
+    /// (k,v,w) ∈ Z
+    /// ```
+    ///
+    /// This is a linear operator.
+    pub fn weigh<F, T>(&self, f: F) -> Stream<Circuit<P>, OrdZSet<Z::Key, T>>
+    where
+        Z: IndexedZSet,
+        Z::Key: Ord + Clone,
+        Z::Val: Ord + Clone,
+        F: Fn(&Z::Key, &Z::Val) -> T + 'static,
+        T: MulByRef<Z::R, Output = T> + MonoidValue,
+    {
+        self.weigh_generic::<_, OrdZSet<_, _>>(f)
+    }
+
+    /// Like [`Self::weigh`], but can return any batch type.
+    pub fn weigh_generic<F, O>(&self, f: F) -> Stream<Circuit<P>, O>
+    where
+        Z: IndexedZSet,
+        Z::Key: Ord + Clone,
+        Z::Val: Ord + Clone,
+        F: Fn(&Z::Key, &Z::Val) -> O::R + 'static,
+        O: Clone + Batch<Key = Z::Key, Val = (), Time = ()> + 'static,
+        O::R: MulByRef<Z::R, Output = O::R>,
+    {
+        self.apply(move |batch| {
+            let mut delta = <O::Builder>::with_capacity((), batch.key_count());
+            let mut cursor = batch.cursor();
+            while cursor.key_valid() {
+                let mut agg = HasZero::zero();
+                while cursor.val_valid() {
+                    agg += f(cursor.key(), cursor.val()).mul_by_ref(&cursor.weight());
+                    cursor.step_val();
+                }
+                delta.push((O::item_from(cursor.key().clone(), ()), agg));
+                cursor.step_key();
+            }
+            delta.done()
+        })
+    }
+}
+
+/// Non-incremental aggregation operator.
+struct Aggregate<Z, A, O> {
+    aggregator: A,
+    _type: PhantomData<(Z, O)>,
+}
+
+impl<Z, A, O> Aggregate<Z, A, O> {
+    pub fn new(aggregator: A) -> Self {
+        Self {
+            aggregator,
+            _type: PhantomData,
+        }
+    }
+}
+
+impl<Z, A, O> Operator for Aggregate<Z, A, O>
+where
+    Z: 'static,
+    A: 'static,
+    O: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("Aggregate")
+    }
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        true
+    }
+}
+
+impl<Z, A, O> UnaryOperator<Z, O> for Aggregate<Z, A, O>
+where
+    Z: IndexedZSet + 'static,
+    Z::Key: PartialEq + Clone,
+    A: Aggregator<Z::Val, (), Z::R> + 'static,
+    O: Clone + IndexedZSet<Key = Z::Key, Val = A::Output> + 'static,
+    O::R: ZRingValue,
+{
+    fn eval(&mut self, i: &Z) -> O {
+        let mut builder = O::Builder::with_capacity((), i.len());
+
+        let mut cursor = i.cursor();
+
+        while cursor.key_valid() {
+            if let Some(agg) = self
+                .aggregator
+                .aggregate(&mut CursorGroup::new(&mut cursor, ()))
+            {
+                builder.push((O::item_from(cursor.key().clone(), agg), O::R::one()));
```

Comment:
Kinda don't like that we're forced to clone the key value for this

Reply:
Hmm, I really don't see how to avoid this.

Reply:
Yah, it was more of a general gripe